base/reflection.jl

```diff
@@ -600,6 +589,29 @@ Return `true` if `x` is an instance of an [`isbitstype`](@ref) type.
 """
 isbits(@nospecialize x) = isbitstype(typeof(x))
 
+"""
+    objectid(x) -> UInt
+
+Get a hash value for `x` based on object identity.
+
+If `x === y` then `objectid(x) == objectid(y)`, and usually when `x !== y`, `objectid(x) != objectid(y)`.
+
+See also [`hash`](@ref), [`IdDict`](@ref).
+"""
+function objectid(x)
+    # objectid is foldable iff it isn't a pointer.
+    # Sring, Symbol, Module, and DataType are safe because jl_objectid for them works by value.
+    if x isa String || Core.Compiler.valid_tparam(x)
```

Comment:
```suggestion
    if isidentityfree(typeof(x))
```
and add test case for e.g. `x::Some{Symbol}`